File: obsidian-mcp-py/obsidian/notes.py
"""
Obsidian Notes Management

Provides tools for creating, reading, updating, listing, and searching notes.
"""

from pathlib import Path
from typing import Optional
from obsidian.vault import get_vault_path, safe_path


def register_note_tools(mcp):
    """
    Register all note-related MCP tools.
    
    Args:
        mcp: FastMCP instance to register tools with
    """
    
    @mcp.tool()
    def read_obsidian_note(note_name: str) -> dict:
        """
        Read a note from Obsidian vault.
        
        Args:
            note_name: Name of the note (without .md extension) or full path relative to vault
        
        Returns:
            Dictionary with note content and metadata
        """
        try:
            vault_path = get_vault_path()
            note_path = safe_path(vault_path, f"{note_name}.md")
            
            # If not found, try as relative path
            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)
            
            if not note_path.exists():
                # Try to find similar note names
                suggestions = []
                try:
                    for md_file in vault_path.rglob("*.md"):
                        if ".obsidian" in md_file.parts:
                            continue
                        if note_name.lower() in md_file.stem.lower():
                            suggestions.append(md_file.stem)
                            if len(suggestions) >= 5:
                                break
                except Exception:
                    pass
                
                return {
                    "error": f"Note not found: {note_name}",
                    "suggestions": suggestions
                }
            
            content = note_path.read_text(encoding="utf-8")
            
            return {
                "note_name": note_path.stem,
                "path": str(note_path.relative_to(vault_path)),
                "content": content,
                "size": len(content),
                "lines": len(content.splitlines())
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def create_obsidian_note(note_name: str, content: str, folder: Optional[str] = None) -> dict:
        """
        Create a new note in Obsidian vault.
        
        Args:
            note_name: Name of the note (without .md extension)
            content: Content of the note in markdown format
            folder: Optional folder path within vault (e.g., "Projects" or "Projects/Tasks")
        
        Returns:
            Dictionary with creation status
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                note_dir = safe_path(vault_path, folder)
                note_dir.mkdir(parents=True, exist_ok=True)
                note_path = note_dir / f"{note_name}.md"
            else:
                note_path = safe_path(vault_path, f"{note_name}.md")
            
            if note_path.exists():
                return {"error": f"Note already exists: {note_name}"}
            
            note_path.write_text(content, encoding="utf-8")
            
            return {
                "success": True,
                "note_name": note_name,
                "path": str(note_path.relative_to(vault_path)),
                "size": len(content)
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def update_obsidian_note(
        note_name: str,
        old_text: Optional[str] = None,
        new_text: Optional[str] = None,
        insert_position: Optional[str] = None,
    ) -> dict:
        """
        Update an existing note in Obsidian vault. Supports partial patch (old_text→new_text) and insert (top/bottom).

        Args:
            note_name: Name of the note (without .md extension) or relative path
            old_text: Text to find and replace (for patch mode)
            new_text: Replacement text (for patch mode) or content to insert
            insert_position: Insert new_text at "top" or "bottom" of note (omit for patch mode)

        Returns:
            Dictionary with update status
        """
        try:
            vault_path = get_vault_path()
            note_path = safe_path(vault_path, f"{note_name}.md")

            if not note_path.exists():
                note_path = safe_path(vault_path, note_name)

            if not note_path.exists():
                return {"error": f"Note not found: {note_name}"}

            existing = note_path.read_text(encoding="utf-8")

            if insert_position:
                if not new_text:
                    return {"error": "new_text is required for insert mode"}
                if insert_position == "top":
                    new_content = new_text + "\n\n" + existing
                elif insert_position == "bottom":
                    new_content = existing + "\n\n" + new_text
                else:
                    return {"error": "insert_position must be 'top' or 'bottom'"}
                mode = f"insert_{insert_position}"
            elif old_text is not None:
                if new_text is None:
                    return {"error": "new_text is required for patch mode"}
                if old_text not in existing:
                    return {"error": "old_text not found in note", "note_name": note_path.stem}
                new_content = existing.replace(old_text, new_text, 1)
                mode = "patch"
            else:
                return {"error": "Provide old_text+new_text (patch) or insert_position+new_text (insert)"}

            note_path.write_text(new_content, encoding="utf-8")
            diff = len(new_content) - len(existing)

            return {
                "success": True,
                "note_name": note_path.stem,
                "path": str(note_path.relative_to(vault_path)),
                "mode": mode,
                "chars_before": len(existing),
                "chars_after": len(new_content),
                "chars_diff": f"+{diff}" if diff >= 0 else str(diff),
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def list_obsidian_notes(folder: Optional[str] = None) -> dict:
        """
        List all notes in Obsidian vault.
        
        Args:
            folder: Optional folder path to list notes from (e.g., "Projects")
        
        Returns:
            Dictionary with list of notes
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "notes": []}
            
            notes = []
            for md_file in search_path.rglob("*.md"):
                # Skip hidden directories and .obsidian folder
                if ".obsidian" in md_file.parts:
                    continue
                
                rel_path = md_file.relative_to(vault_path)
                notes.append({
                    "name": md_file.stem,
                    "path": str(rel_path),
                    "size": md_file.stat().st_size
                })
            
            notes.sort(key=lambda x: x["path"])
            
            return {
                "vault_path": str(vault_path),
                "folder": folder or "root",
                "total_notes": len(notes),
                "notes": notes
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def search_obsidian_notes(query: str, folder: Optional[str] = None) -> dict:
        """
        Search for notes containing specific text in Obsidian vault.
        
        Args:
            query: Text to search for in note content
            folder: Optional folder path to limit search (e.g., "Projects")
        
        Returns:
            Dictionary with matching notes
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "matches": []}
            
            matches = []
            query_lower = query.lower()
            
            for md_file in search_path.rglob("*.md"):
                # Skip hidden directories and .obsidian folder
                if ".obsidian" in md_file.parts:
                    continue
                
                try:
                    content = md_file.read_text(encoding="utf-8")
                    if query_lower in content.lower():
                        rel_path = md_file.relative_to(vault_path)
                        # Find line numbers with matches
                        lines = content.splitlines()
                        matching_lines = [
                            i + 1 for i, line in enumerate(lines)
                            if query_lower in line.lower()
                        ]
                        
                        matches.append({
                            "name": md_file.stem,
                            "path": str(rel_path),
                            "matching_lines": matching_lines[:5],  # First 5 matches
                            "total_matches": len(matching_lines)
                        })
                except Exception:
                    continue
            
            matches.sort(key=lambda x: x["total_matches"], reverse=True)
            
            return {
                "query": query,
                "folder": folder or "root",
                "total_matches": len(matches),
                "matches": matches
            }
        except Exception as e:
            return {"error": str(e)}

File: obsidian-mcp-ts/src/obsidian/notes.ts
import fs from "fs";
import path from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getVaultPath } from "./vault.js";
import { findMarkdownFiles, resolveNotePath, relPath, stem, safePath } from "./utils.js";

export function registerNoteTools(server: McpServer) {
  server.tool(
    "read_obsidian_note",
    "Read a note from Obsidian vault",
    { note_name: z.string().describe("Name of the note (without .md extension) or full path relative to vault") },
    async ({ note_name }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);

        if (!notePath) {
          const suggestions: string[] = [];
          for (const mdFile of findMarkdownFiles(vaultPath)) {
            if (stem(mdFile).toLowerCase().includes(note_name.toLowerCase())) {
              suggestions.push(stem(mdFile));
              if (suggestions.length >= 5) break;
            }
          }
          return { content: [{ type: "text", text: JSON.stringify({ error: `Note not found: ${note_name}`, suggestions }) }] };
        }

        const content = fs.readFileSync(notePath, "utf-8");
        return {
          content: [{
            type: "text",
            text: JSON.stringify({
              note_name: stem(notePath),
              path: relPath(vaultPath, notePath),
              content,
              size: content.length,
              lines: content.split("\n").length,
            }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text", text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "create_obsidian_note",
    "Create a new note in Obsidian vault",
    {
      note_name: z.string().describe("Name of the note (without .md extension)"),
      content: z.string().describe("Content of the note in markdown format"),
      folder: z.string().optional().describe("Optional folder path within vault"),
    },
    async ({ note_name, content, folder }) => {
      try {
        const vaultPath = getVaultPath();
        let notePath: string;

        if (folder) {
          const noteDir = safePath(vaultPath, folder);
          fs.mkdirSync(noteDir, { recursive: true });
          notePath = path.join(noteDir, `${note_name}.md`);
        } else {
          notePath = safePath(vaultPath, `${note_name}.md`);
        }

        if (fs.existsSync(notePath)) {
          return { content: [{ type: "text", text: JSON.stringify({ error: `Note already exists: ${note_name}` }) }] };
        }

        fs.writeFileSync(notePath, content, "utf-8");
        return {
          content: [{
            type: "text",
            text: JSON.stringify({ success: true, note_name, path: relPath(vaultPath, notePath), size: content.length }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text", text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "update_obsidian_note",
    "Update an existing note in Obsidian vault. Supports partial patch (old_text→new_text) and insert (top/bottom).",
    {
      note_name: z.string().describe("Name of the note (without .md extension) or relative path"),
      old_text: z.string().optional().describe("Text to find and replace (for patch mode)"),
      new_text: z.string().optional().describe("Replacement text (for patch mode) or content to insert"),
      insert_position: z.enum(["top", "bottom"]).optional().describe("Insert new_text at top or bottom of note (omit for patch mode)"),
    },
    async ({ note_name, old_text, new_text, insert_position }) => {
      try {
        const vaultPath = getVaultPath();
        const notePath = resolveNotePath(vaultPath, note_name);

        if (!notePath) {
          return { content: [{ type: "text", text: JSON.stringify({ error: `Note not found: ${note_name}` }) }] };
        }

        const existing = fs.readFileSync(notePath, "utf-8");
        let newContent: string;
        let mode: string;

        if (insert_position) {
          // Insert mode: add new_text at top or bottom
          if (!new_text) {
            return { content: [{ type: "text", text: JSON.stringify({ error: "new_text is required for insert mode" }) }] };
          }
          newContent = insert_position === "top"
            ? new_text + "\n\n" + existing
            : existing + "\n\n" + new_text;
          mode = `insert_${insert_position}`;
        } else if (old_text !== undefined) {
          // Patch mode: find and replace
          if (new_text === undefined) {
            return { content: [{ type: "text", text: JSON.stringify({ error: "new_text is required for patch mode" }) }] };
          }
          if (!existing.includes(old_text)) {
            return { content: [{ type: "text", text: JSON.stringify({ error: "old_text not found in note", note_name: stem(notePath) }) }] };
          }
          newContent = existing.replace(old_text, new_text);
          mode = "patch";
        } else {
          return { content: [{ type: "text", text: JSON.stringify({ error: "Provide old_text+new_text (patch) or insert_position+new_text (insert)" }) }] };
        }

        fs.writeFileSync(notePath, newContent, "utf-8");
        const diff = newContent.length - existing.length;
        return {
          content: [{
            type: "text",
            text: JSON.stringify({
              success: true,
              note_name: stem(notePath),
              path: relPath(vaultPath, notePath),
              mode,
              chars_before: existing.length,
              chars_after: newContent.length,
              chars_diff: diff >= 0 ? `+${diff}` : `${diff}`,
            }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text", text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "list_obsidian_notes",
    "List all notes in Obsidian vault",
    { folder: z.string().optional().describe("Optional folder path to list notes from") },
    async ({ folder }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? safePath(vaultPath, folder) : vaultPath;

        if (!fs.existsSync(searchPath)) {
          return { content: [{ type: "text", text: JSON.stringify({ error: `Folder not found: ${folder}`, notes: [] }) }] };
        }

        const notes = findMarkdownFiles(searchPath).map((mdFile) => ({
          name: stem(mdFile),
          path: relPath(vaultPath, mdFile),
          size: fs.statSync(mdFile).size,
        }));

        notes.sort((a, b) => a.path.localeCompare(b.path));

        return {
          content: [{
            type: "text",
            text: JSON.stringify({ vault_path: vaultPath, folder: folder || "root", total_notes: notes.length, notes }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text", text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "search_obsidian_notes",
    "Search for notes containing specific text",
    {
      query: z.string().describe("Text to search for in note content"),
      folder: z.string().optional().describe("Optional folder path to limit search"),
    },
    async ({ query, folder }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? safePath(vaultPath, folder) : vaultPath;

        if (!fs.existsSync(searchPath)) {
          return { content: [{ type: "text", text: JSON.stringify({ error: `Folder not found: ${folder}`, matches: [] }) }] };
        }

        const queryLower = query.toLowerCase();
        const matches: any[] = [];

        for (const mdFile of findMarkdownFiles(searchPath)) {
          try {
            const content = fs.readFileSync(mdFile, "utf-8");
            if (content.toLowerCase().includes(queryLower)) {
              const lines = content.split("\n");
              const matchingLines = lines
                .map((line, i) => (line.toLowerCase().includes(queryLower) ? i + 1 : -1))
                .filter((n) => n !== -1);

              matches.push({
                name: stem(mdFile),
                path: relPath(vaultPath, mdFile),
                matching_lines: matchingLines.slice(0, 5),
                total_matches: matchingLines.length,
              });
            }
          } catch {
            continue;
          }
        }

        matches.sort((a, b) => b.total_matches - a.total_matches);

        return {
          content: [{
            type: "text",
            text: JSON.stringify({ query, folder: folder || "root", total_matches: matches.length, matches }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text", text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );
}
